whylog/config/__init__.py

```diff
@@ -51,57 +59,68 @@ def _save_rule_definition(self, rule_definition):
-        earliest_cause_time = datetime(2015, 12, 3, 12, 8, 8)
-        default_investigation_step = InvestigationStep(
-            concatenated, effect_time, earliest_cause_time
-        )
-        rule = Rule(
-            [cause], effect, [
-                {
-                    'clues_groups': [[1, 1], [0, 1]],
-                    'name': 'time',
-                    'params': {'max_delta': 1}
-                }
-            ]
-        )  # yapf: disable
         line_source = LineSource('localhost', 'node_1.log', 40)
         effect_clues = {'effect': Clue((effect_time,), 'node_1.log', line_source)}
-        return InvestigationPlan(
-            [rule], [(default_investigation_step, default_log_type)], effect_clues
+        investigation_data = self._create_steps_in_investigation(
+            concatenated_parsers, suspected_rules, effect_clues
         )
-
-    def create_investigation_plan(self, front_input, log_type_name):
-        # TODO: remove mock
-        return self.mocked_investigation_plan()
+        return InvestigationPlan(suspected_rules, investigation_data, effect_clues)
 
     def get_log_type(self, front_input):
         # TODO: remove mock
         matcher = RegexFilenameMatcher('localhost', 'node_1.log', 'default')
         return LogType('default', [matcher])
 
-    def _find_matching_parsers(self, front_input, log_type):
-        pass
+    def _find_matching_parsers(self, front_input, log_type_name):
+        matching_parsers = []
+        extracted_params = {}
+        for parser in self._parsers_grouped_by_log_type[log_type_name]:
+            params = parser.get_regex_params(front_input.line_content)
+            if params is not None:
+                extracted_params[parser.name] = params
+                matching_parsers.append(parser)
+        return matching_parsers, extracted_params
 
     def _filter_rule_set(self, parsers_list):
-        pass
+        suspected_rules = []
+        for parser in parsers_list:
+            if self._rules.get(parser.name) is not None:
+                suspected_rules.extend(self._rules.get(parser.name))
+        return suspected_rules
+
+    @classmethod
+    def _create_concatenated_parser_for_investigation(cls, rules):
+        grouped_parsers = defaultdict(list)
+        inserted_parsers = set([])
+        for suspected_rule in rules:
+            for parser in suspected_rule.get_causes_parsers():
+                if parser.name not in inserted_parsers:
+                    grouped_parsers[parser.log_type].append(parser)
+                    inserted_parsers.add(parser.name)
+        return dict(
+            (log_type_name, ConcatenatedRegexParser(parsers))
+            for log_type_name, parsers in grouped_parsers.items()
+        )
 
-    def _get_locations_for_logs(self, logs_types_list):
-        pass
+    def _create_steps_in_investigation(self, concatenated_parsers, suspected_rules, effect_clues):
+        investigation_data = []
```

Comment:
I'm not sure if `investigation_data` is a good choice for this variable name. Everything around is data, and naming something `data` doesn't giving any help. I see that `investigation_data` is a list consising of pairs: `investigation_step, log_type`. Maybe `steps_and_log_types` will be better? Or something like that.